fragment.go

```diff
@@ -2266,6 +2450,59 @@ func (f *fragment) rows(start uint64, filters ...rowFilter) []uint64 {
 	return rows
 }
 
+// upgradeRoaringBSIv2 upgrades a fragment that contains old BSI formatting
+// to a new BSI format (v2). The new format moves the "exists" bit to the
+// beginning & adds a negative sign bit.
+func upgradeRoaringBSIv2(f *fragment, bitDepth uint) (string, error) {
+	// If flag set, already upgraded. Exit.
+	if f.storage.Flags&roaringFlagBSIv2 == 1 {
+		return "", nil
+	}
+
+	other := roaring.NewBitmap()
+	other.Flags = roaringFlagBSIv2
+	if err := func() error {
+		f.mu.Lock()
+		defer f.mu.Unlock()
+
+		var err error
```

Comment:
I like that this seems reasonably clear, but I think it's probably going to be fairly slow. Idle thought: We could move the *containers* rather than the *bits*. Each container present in the old system has a correct position in the new system; if it's the existence bit, it's reduced by ShardWidth*bitDepth, otherwise it's increased by 2*ShardWidth. If we have a lock on the storage (so we know no one's modifying it until we're done), we don't even need to clone the containers, we can just insert them in the new storage file at the correct positions.

Reply:
As long as it's not going to take, like, hours... I don't think we should spend much effort on optimizing the upgrade code since it isn't going to run over and over.

Reply:
that's probably fair. i don't actually have a clear sense of how large the data sets are or how long it'll take.